### PR TITLE
[20841] Move custom fields settings to its own tab

### PR DESCRIPTION
**Pull request by basiszwo**
https://community.openproject.org/work_packages/20841

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://basiszwo-20841-custom-fields.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://basiszwo-20841-custom-fields.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by basiszwo** [comment]
@ulferts so here's my first contribution. Looking forward to feedback :smiley:

File: app/helpers/projects_helper.rb [hunk 1]
```diff
@@ -39,6 +39,7 @@ def project_settings_tabs
     tabs = [{ name: 'info', action: :edit_project, partial: 'projects/edit', label: :label_information_plural },
             { name: 'modules', action: :select_project_modules, partial: 'projects/settings/modules', label: :label_module_plural },
             { name: 'members', action: :manage_members, partial: 'projects/settings/members', label: :label_member_plural },
+            { name: 'custom_fields', action: :edit_project, partial: 'projects/settings/custom_fields', label: :label_custom_field_plural },
```
Comment (floriank) [inline app/helpers/projects_helper.rb:42]: uuuh, them hound critics. such code style. much wow.

image

:wink:

File: app/views/projects/settings/_custom_fields.html.erb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+See doc/COPYRIGHT.rdoc for more details.
+
+++#%>
+
+<%= labelled_tabular_form_for @project, :method => :put,
```
Comment (floriank) [inline app/views/projects/settings/_custom_fields.html.erb:30]: We should really go for the JSON-esque syntax of `method: :put`.

**Comment by floriank** [comment]
~~I would suspect that there either are integration tests that should be in place or will break.~~

EDIT: apparently i am too lazy to read the full commit.

also, this should probably be merged into `release/4.3`

**Comment by basiszwo** [comment]
wasn't sure about the base to create the PR against. will update code and create a new PR.

Have you seen the changed cuke?

File: app/views/projects/settings/_custom_fields.html.erb (hunk 2 above)
Comment (NobodysNightmare) [inline app/views/projects/settings/_custom_fields.html.erb:30]: I am irritated that the @houndci did not find that offense...

Comment (floriank) [inline app/views/projects/settings/_custom_fields.html.erb:30]: It's a template

**Comment by floriank** [comment]
> wasn't sure about the base to create the PR against. will update code and create a new PR.

It's a small PR so it will not matter, we can merge this manually against `release/4.3` and merge it back to `dev`. ultimately it will live in both branches.

The indicator is the version in the work package.

**Comment by NobodysNightmare** [comment]
> It's a small PR so it will not matter, we can merge this manually against `release/4.3` and merge it back to `dev`.

be careful, merge errors happen to the best of us they say :wink:

File: app/views/projects/settings/_custom_fields.html.erb (hunk 2 above)
Comment (myabc) [inline app/views/projects/settings/_custom_fields.html.erb:30]: Unfortunately rubocop doesn't support ERB.

**Comment by floriank** [comment]
> be careful, merge errors happen to the best of us they say :wink: 

and to you, as well :wink: !

**Comment by ulferts** [comment]
@basiszwo this PR seems to do more than it says:
- Remove heading on the modules tab
- Remove fieldset on types tab (but leaves heading)
- changes the layout of custom fields (seems to be covered by the work package)

If the headings are to be removed on all tabs, and it would make them consistent to e.g. the administrative settings, then the keys in [confing/locales/en.yml](https://github.com/opf/openproject/blob/dev/config/locales/en.yml) should be removed as well.

**Comment by ulferts** [comment]
@basiszwo the form on the custom fields tab is sent via PUT to `/projects/custom_fields/[project identifier]`. This only works because we have a [default routes](https://github.com/opf/openproject/blob/dev/config/routes.rb#L582) configured which we want to get rid of. The form should be sent to `/projects/[project identifier]/custom_fields`. Please define an appropriate route specifically just like it is done for [modules](https://github.com/opf/openproject/blob/dev/config/routes.rb#L236). Please also add tests for the route. There are a lot of [examples](https://github.com/opf/openproject/blob/dev/spec/routing/journals_spec.rb) you can rely on.

File: app/controllers/projects_controller.rb [hunk 3]
```diff
@@ -183,6 +183,13 @@ def modules
     redirect_to action: 'settings', id: @project, tab: 'modules'
   end
 
+  def custom_fields
+    @project.update_attributes(params[:project])
+    flash[:notice] = l(:notice_successful_update)
+
+    redirect_to action: 'settings', id: @project, tab: 'custom_fields'
```
Comment (ulferts) [inline app/controllers/projects_controller.rb:190]: - The method does not enforce authorization. Please add the action to [the ones requiring authorization](https://github.com/basiszwo/openproject/blob/20841-custom-fields/app/controllers/projects_controller.rb#L39). 
- There are no specs for this action
- The action does allow the user to alter more than just setting the custom_fields of the project. I could e.g. change the name of the project using the method by manipulating the form. You should either rely on [permitted params](https://github.com/basiszwo/openproject/blob/dev/app/models/permitted_params.rb) or you could take the desired values out of the hash yourself. As long as the permissions are related, this is not a problem. But if we introduce a separate permission, the user will be allowed to do more than anticipated.
- There is no error handling in the method. If I manipulate the form to send a non existing ID e.g. -1, an ActiveRecord error will be thrown. (granted, this weakness also exists in the `modules` method above.

File: app/views/projects/form/_custom_fields.html.erb [hunk 4]
```diff
@@ -28,17 +28,8 @@ See doc/COPYRIGHT.rdoc for more details.
...
 <fieldset class="form--fieldset -with-control" id="project_issue_custom_fields">
-  <legend class="form--fieldset-legend"><%=l(:label_custom_field_plural)%></legend>
-  <div class="form--fieldset-control">
-    <span class="form--fieldset-control-container">
-      (<%= check_all_links 'project_issue_custom_fields' %>)
-    </span>
-  </div>
-
-  <ul class="form--list">
     <% work_package_custom_fields.each do |custom_field| %>
```
Comment (ulferts) [inline app/views/projects/form/_custom_fields.html.erb:31]: indentation is off, please move to the left.

**Comment by ulferts** [comment]
@basiszwo by removing the headings, the create form does look weird now. I propose moving the headings to the form then, so that the form is more pleasing to look at again.

Additionally, by removing the custom fields from the form, they are no longer part of the create form. I don't know if that is intentional.

File: features/projects/settings.feature [hunk 5]
```diff
@@ -86,7 +86,7 @@ Feature: Project Settings
     When the following issue custom fields are defined:
       | name             | type      | is_for_all |
       | My Custom Field  | text      | false      |
-    And I go to the settings page of the project "project1"
-    And I check "My Custom Field" within "#tab-content-info"
-    And I press "Save" within "#tab-content-info"
+    And I go to the custom_fields tab of the settings page of the project "project1"
+    And I check "My Custom Field" within "#tab-content-custom_fields"
+    And I press "Save" within "#tab-content-custom_fields"
     Then the "My Custom Field" checkbox should be checked
```
Comment (ulferts) [inline features/projects/settings.feature:92]: @basiszwo you can remove the `within ...` part now that this here is on it's own tab. And this does not require selenium to be run. So while you are touching the code, please remove the `@javascript` statement from this scenario. This speeds up test execution.

**Comment by basiszwo** [comment]
Will reimplement this against current dev.

**Pull request closed without merging**